### ISSUE TITLE
[CI] RareClusterStateIT.testDeleteCreateInOneBulk

### ISSUE THREAD

**Issue by vladimirdolzhenko** [elastic/elasticsearch#30718]
log: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+intake/1928/console

```
  2> REPRODUCE WITH: ./gradlew :server:integTest -Dtests.seed=50A45DB571AFB30D -Dtests.class=org.elasticsearch.indices.state.RareClusterStateIT -Dtests.method="testDeleteCreateInOneBulk" -Dtests.security.manager=true -Dtests.locale=pt-PT -Dtests.timezone=Asia/Aqtobe
ERROR   6.92s J1 | RareClusterStateIT.testDeleteCreateInOneBulk <<< FAILURES!
   > Throwable #1: Failed to execute phase [query], all shards failed
   > 	at __randomizedtesting.SeedInfo.seed([50A45DB571AFB30D:C49B930FC733BE6B]:0)
   > 	at org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseFailure(AbstractSearchAsyncAction.java:293)
   > 	at org.elasticsearch.action.search.AbstractSearchAsyncAction.executeNextPhase(AbstractSearchAsyncAction.java:133)
   > 	at org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseDone(AbstractSearchAsyncAction.java:254)
   > 	at org.elasticsearch.action.search.InitialSearchPhase.onShardFailure(InitialSearchPhase.java:101)
   > 	at org.elasticsearch.action.search.InitialSearchPhase.access$100(InitialSearchPhase.java:48)
   > 	at org.elasticsearch.action.search.InitialSearchPhase$2.lambda$onFailure$1(InitialSearchPhase.java:222)
   > 	at org.elasticsearch.action.search.InitialSearchPhase.maybeFork(InitialSearchPhase.java:176)
   > 	at org.elasticsearch.action.search.InitialSearchPhase.access$000(InitialSearchPhase.java:48)
   > 	at org.elasticsearch.action.search.InitialSearchPhase$2.onFailure(InitialSearchPhase.java:222)
   > 	at org.elasticsearch.action.search.SearchExecutionStatsCollector.onFailure(SearchExecutionStatsCollector.java:73)
   > 	at org.elasticsearch.action.ActionListenerResponseHandler.handleException(ActionListenerResponseHandler.java:51)
   > 	at org.elasticsearch.action.search.SearchTransportService$ConnectionCountingHandler.handleException(SearchTransportService.java:535)
   > 	at org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleException(TransportService.java:1095)
   > 	at org.elasticsearch.transport.TransportService$DirectResponseChannel.processException(TransportService.java:1188)
   > 	at org.elasticsearch.transport.TransportService$DirectResponseChannel.sendResponse(TransportService.java:1172)
   > 	at org.elasticsearch.transport.TaskTransportChannel.sendResponse(TaskTransportChannel.java:66)
   > 	at org.elasticsearch.action.search.SearchTransportService$6$1.onFailure(SearchTransportService.java:393)
   > 	at org.elasticsearch.search.SearchService$2.onFailure(SearchService.java:340)
   > 	at org.elasticsearch.search.SearchService$2.onResponse(SearchService.java:334)
   > 	at org.elasticsearch.search.SearchService$2.onResponse(SearchService.java:328)
   > 	at org.elasticsearch.search.SearchService$3.doRun(SearchService.java:1015)
   > 	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:724)
   > 	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
   > 	at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:41)
   > 	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
   > 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
   > 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
   > 	at java.lang.Thread.run(Thread.java:748)
```

most probably it is due to last line of test 
```
assertHitCount(client().prepareSearch("test").get(), 0);
```

@elastic/es-distributed could somebody have a look into that ?

**Comment by bleskes** [comment]
@ywelsch do you mind taking a look?

**Comment by tlrx** [comment]
Another instance here: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+periodic/6349/consoleFull

[consoleText.txt](https://github.com/elastic/elasticsearch/files/2026194/consoleText.txt)

**Comment by ywelsch** [comment]
The test has an issue that exhibits only super rarely. The test sets the publish timeout to 0, then proceeds to block cluster state processing on a data node, then deletes an index and recreates it, and finally removes the cluster state processing block. Finally, it calls ensureGreen, which might now return before the data node has fully applied the cluster state that removed and readded the shard, due to the publish timeout of 0. I've pushed a fix (e6a784c4748) to the test that waits for the cluster state to be fully processed on the data node before doing the search.
